include/dqlite.h

```diff
@@ -187,7 +187,18 @@ struct dqlite_node_info
 };
 typedef struct dqlite_node_info dqlite_node_info;
 
+struct dqlite_node_info_v2
+{
+	dqlite_node_id id;
+	const char *address;
+	int dqlite_role;
+};
+typedef struct dqlite_node_info_v2 dqlite_node_info_v2;
```

Comment:
Can't you add the field to the existing `dqlite_node_info` struct and put a size check or similar in place to detect someone passing you the older struct?

I think @brauner did a bunch of that kind of stuff in liblxc where we need to expand structs and don't want to needlessly have to fork them and duplicate symbols.

Reply:
We'll want @brauner to comment here. For liblxc I believe it's a mix of memset and some checks so if we're fed a smaller struct, we're not reading garbage and can tell if the data was provided or not.

If this isn't doable with the current symbol here, I'd at least argue that the new symbol we're adding should allow for further changes to the struct without then needing a v3.

Reply:
It is possible to version a C struct by size and is pretty common. Aleksa and I introduced a dedicated API for this into the kernel a while back. For this to work nicely the struct should usually be aligned at a 64 bit boundary. That ensures that the compiler doesn't add padding. I'm quoting from the document a document we wrote back then:

```rst
Advanced extensibility: extensible structs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For more sophisticated system calls that involve a larger number of arguments,
it's preferred to encapsulate the majority of the arguments into an extensible
structure that is passed in by pointer.

Extensible structs are versioned by their size.  For any new non-flag based
extension a new field has to be added to the end of the extensible struct.  The
zero value of the new field must not have any meaning so the system call can
continue to display the old behavior.

Extensible struct system calls can and should use the dedicated
``copy_struct_from_user`` API which enforces the following behavior:

 - The kernel will reject any size that is smaller than the initially supported
   size of the extensible struct.
 - If the extensible struct size the kernel knows about is equal to the size
   passed in from userspace then ``copy_struct_from_user`` will copy the struct
   verbatim.
 - If the extensible struct size the kernel knows about is larger than the size
   passed in from userspace the kernel will copy the size userspace indicated
   and treat all additional extensions it knows about as zero.
 - If the extensible struct size the kernel knows about is smaller than the
   size passed in from userspace the kernel will copy the number of bytes it
   knows about and verify that all trailing bytes are zero.  If non-zero bytes
   are present the kernel returns ``E2BIG``.  While not an intuitive error code
   at first, ``E2BIG`` means that the argument list is too long.

Early examples for extensible struct system calls include
:manpage:`perf_event_open(2)` and :manpage:`sched_setattr(2)`.  These system
calls implement mostly similar behavior even before the introduction of
``copy_struct_from_user`` but have since been switched over to it.  Newer
examples include :manpage:`clone3(2)` and :manpage:`openat2(2)`.

The size associated with an extensible struct can either be passed as
a separate argument in the system call as is e.g. done for :manpage:`clone3(2)`
and :manpage:`openat2(2)`.  Alternatively, the size can be passed as the first
field in the extensible struct as is e.g. done for :manpage:`sched_setattr(2)`.

Any struct passed from userspace to the kernel and especially extensible
structs must ensure that they are correctly padded.  This ensures that no data
can be leaked on accident or on purpose by an attacker from the kernel.  The
easiest way to ensure that a struct is correctly padded is to always use 64 bit
fields::

    struct sys_foo_args {
        __aligned_u64 arg1;
        __aligned_u64 arg2;
        __aligned_u64 arg3;
        __aligned_u64 arg4;
        __aligned_u64 arg5;
    };

System calls that need to worry about the size of their extensible structs or
need fields to be of a specific size can rely on careful manual struct
packing::

    struct sys_foo_args {
        __u32 arg1;
        __u16 arg2;
        __u16 arg3;
        __u32 arg4;
        __u32 arg5;
        __u64 arg6;
    };

(There are tools such as ``pahole`` available that allow to check whether
a struct is correctly padded!)

Note that in contrast to flag arguments passed as register arguments flag
arguments in extensible structures can be 64 bit wide.  As with simple
flag-only system calls, the system call needs to verify any unknown values for
flag-like fields in the passed struct are zeroed.
```